### PR TITLE
Mario offsetcalculation

### PR DESCRIPTION
**Pull request by Froehliche-Kernschmelze**
code is ready for a group review. 

contains: 
- the entire offset calculation, stored in /offsets/ as .csv files
- a function that applies the offset calculation, reading the .csv files of the above mentioned directory, and returns an array of calibrated data

---

Needed:
- [x] resolve conflicts with master
- [x] remove most of the comments, e.g. "explain the line before", "explain syntax"
- [x] remove plotting code entirely from the calculation code
- [x] provide implementation to also read the offsets in, after having them written to files. Writer/Reader pair.
- [x] remove print statements. A program that works should mostly be quiet.
- [x] implement the `offset_apply` wrapper, as proposed by @MaxNoe, see #11 
- [x] remove "magic numbers", i.e. 4096 and 1000.
- [x] remove "magic strings", i.e. "low" and "high". 
- [x] take apart "calculation" and "storage" of calibration constants.

Nice to have:
 - [x] add __doc_strings__ where needed.
 - [ ] Loop through each file only once.
 - [x] Do not assume something about the Filename of the Pedestal files.
 - [x] Store the offset constants in a single file only, e.g. HDF5, shelve, pickle, npz(?) ...

### CODE REVIEW

**Comment by MaxNoe** [comment]
A few thoughts:

* the results should not be pushed to the repo. These are large files, which can be reproduced quickly. They do not belong into the VCS imho.

* use `os.path.join` to build paths, do not manually to stuff like `path` + '\' + 'path2`

* In the current implemenation, you read the file again for every channel. This is not needed. Loop through the file once, use 16 `RunningStats` objects to fill in the data for each pixel. This might be done in a loop like this:


```
# initialize dict with Runningstatsobjects for each channel
stats = {pixelindex, gain_channel: RunningStats(4096)
            for pixelindex, gain_channel in zip(range(8), ['high', 'low')}
```  

And then fill it:
```
for pixel_index in range(8):
    for gain in ['low', 'high']:
        # do stuff 
        stats[pixel_index, gain].fill(stuff)
```

* the main function currently does not do the offset_calibration, this is misleading I think.

**Comment by dneise** [comment]
>>> the results should not be pushed to the repo. These are large files, which can be reproduced quickly. They do not belong into the VCS imho.

compared to the example file "Ped444706_1.dat" --> 32MB the size of the calibration files --> 3MB is not worth discussing.

Also, being able to plot calibrated example events using the example file and the offsets, directly after cloning the repo is a nice feature.

So I vote for leaving the stored offsets in the repo, at least for a while.. once we get minicam data we should certainly re-discuss.

**Comment by dneise** [comment]
I support all the rest what Max said

**Comment by dneise** [comment]
Regarding the application of the constants, I already wrote somewhere else, that I think, the *File* or *EventGetter*, should take care of it. It is imho not necessary to have a stand-alone executable, which converts raw-data to calibrated-data or so...

**Comment by dneise** [comment]
Also I think saving all calibration constants in a structure similar to `event.data` makes sense.
So say, I want the offsets from channel 2, low gain, cells 1200:1600, I would like to write:

```python
my_offsets = offsets['low'][2][1200:1600]
```
Or something like that.

**Comment by dneise** [comment]
Maybe, we should go and use a default path like `~/.dragonboard/Boardname_offsets.foo`
or  `~/.dragonboard/Boardname/offsets.foo` to store the dragonboard offsets, when they are calculated.

Then we teach the dragonviewer to look there for a calibration file. 

Also the `File` or `EventGetter` can look there, so one can do something like:

```python

import dragonboard as dbd
with dbd.EventGetter("some_path/to/raw_file.dat", calibrate=True) as f:
   plt.plot(f[0].data["low"][2])
```

**Comment by MaxNoe** [comment]
Please see #15.

Also I would suggest to get rid of most of the more extensive comments, for example at the imports.

Also we should start using the standard docstrings described here: https://www.python.org/dev/peps/pep-0257/

This will make it easy to setup a sphinx documentation when this becomes bigger.

File: offset_calculation.py [hunk 1]
```diff
@@ -1,38 +1,107 @@
...
+from tqdm import tqdm # enable to show progress
+import glob # enable to search for files in directories
+import os
+from docopt import docopt # write your own menu! furthermore control terminal input parameters
+
+
+def offset_calc(filedirectory, outputdirectory, pixelindex, gaintype):
+    """ calculate mean offset & RMS for every capacitor and plot the data. the data is saved as .csv files.
+        The .csv files are named offsets_channel<pixelindex>_<gaintype>-gain. 
+        this FORMAT MUST NOT BE CHANGED
```
Comment (MaxNoe) [inline offset_calculation.py]: Really? Must not be changed? At this early stage?

File: offset_calculation.py [hunk 2]
```diff
@@ -1,38 +1,107 @@
...
+    """
+
+    stats = RunningStats(shape=4096) # initialize stats array on which calculations are carried out
+
+    for filename in glob.glob(os.path.join(filedirectory, 'Ped*.dat')): # search for all .dat's in the current directory
+
+        with open(filename, "rb") as f: # open a file and call it f. "rb": read, file is binary
+
+            max_events = 1000 # 1000 at all
```
Comment (MaxNoe) [inline offset_calculation.py]: Why `max_events = 1000`, in general, we want to use them all, right? Better just leave it out.

File: offset_calculation.py [hunk 3]
```diff
@@ -1,38 +1,107 @@
...
+    plt.errorbar(
+        np.arange(4096),
+        stats.mean,
+        yerr=stats.std, # yerr = y-error bars
+        fmt="+", # fmt means format
+        markersize=3,
+        capsize=1, # frame bars of error bars
+    )
+    plt.xlim(0,4096)
+    plt.figure()
```
Comment (MaxNoe) [inline offset_calculation.py]: The `plt.figure` at the bottom results in an empty figure. Put it on top of the plot commands.

**Comment by MaxNoe** [comment]
Oh. And I would remove the plotting code from this script, or at least make it optional with an `--plot` command line option.

**Comment by dneise** [comment]
I would certainly remove the plotting code from a function called "offset_calc". A function should be a tiny thing, doing exactly what is seems to do judging from its name. 

When you seem to be unable to give it a proper name, then this is a hint, that a function is maybe doing too many different things.

As Max said: using a `--plot` cmd line argument is a way to go .. but also ... having in mind that plotting a couple of numbers, which are stored in an array is a one-liner. Why providing any code for it.
Who wants to see that plot, needs to type one line (after reading the offsets)
plt.plot(offsets['low'][2])

and thus can make the plot in the color, and style he or she prefers... No need for us to provide plotting code here. What we *should* provide imho is some easy means of reading the offsets into a nice array

**Comment by MaxNoe** [comment]
@dneise I completly agree

**Comment by mackaiver** [comment]
@Froehliche-Kernschmelze Great work. I think we can merge this after you fix the few minor things mentioned above. Nicely done.

**Comment by dneise** [comment]
So let's clearly define these minor things, that need to be done, so @Froehliche-Kernschmelze  knows what needs to be done.

Needed:
- [ ] resolve conflicts with master
- [ ] remove most of the comments, e.g. "explain the line before", "explain syntax"
- [ ] remove plotting code entirely from the calculation code
- [ ] provide implementation to also read the offsets in, after having them written to files. Writer/Reader pair.
- [ ] remove print statements. A program that works should mostly be quiet.
- [ ] implement the `offset_apply` wrapper, as proposed by @MaxNoe 

Nice to have:
 - [ ] add __doc_strings__ where needed.
 - [ ] Loop through each file only once.

(Can people edit this, or can only I do that?)

**Comment by MaxNoe** [comment]
Only you. But you can put it into the description.

**Comment by dneise** [comment]
perfect

**Comment by Froehliche-Kernschmelze** [comment]
I don't know if this is  the right place to post this. After a pull from this repo my (I think unchanged) my offset_calculation.py code throws some errors. I know what it is saying but I am not sure why the code has worked out fine before with the same configs and does not now. Does anyone have a clue why?

python offset_calculation.py /home/mario/Programs/IP192.168.1.1/Pedestal/ /home/mario/Programs/lst_calibration/ 

0it [00:00, ?it/s]/home/mario/Programs/lst_calibration/dragonboard/runningstats.py:19: VisibleDeprecationWarning: boolean index did not match indexed array along dimension 0; dimension is 1024 but corresponding boolean dimension is 4096
  self._M2 = self._M2[idx] + self._delta[idx] * (data[idx] - self._mean[idx])
Traceback (most recent call last):
  File "offset_calculation.py", line 124, in <module>
    store_data(outputdirectory, offset_calc(inputdirectory))
  File "offset_calculation.py", line 65, in offset_calc
    stats.add(np.roll(data, stop_cell[stop_cell_array_pos])) # STATS.ADD causes problems :-(
  File "/home/mario/Programs/lst_calibration/dragonboard/runningstats.py", line 19, in add
    self._M2 = self._M2[idx] + self._delta[idx] * (data[idx] - self._mean[idx])
IndexError: index 2626 is out of bounds for axis 1 with size 1024

**Comment by Froehliche-Kernschmelze** [comment]
By the way, any feedback regarding offset_calibration.py would be awesome! code works all nice and quick although it lacks some speed while writing the calibrated data to a file.

**Comment by MaxNoe** [comment]
I fixed the error in `runningstats`, again. I know that I had fixed this already, dunno how this got lost. Maybe in this cleaning madness.

**Comment by dneise** [comment]
When invoking `offset_calculation.py` I see many many `tqdm` bars flicking into live and disappear faster than I can look at them. IMO this makes no sense.

**Comment by dneise** [comment]
We do not comment out large parts of code... and Write `#Junk` on top of it. We delete that.
If we need it again, its still somewhere in the history.

**Comment by dneise** [comment]
When calling `offset_calculation.py` like this
```
time python offset_calculation.py ../IP192.168.1.1/Pedestal/ ./constants
```

It does not create the output-folder `./constants`, but it creates an outputfile called `offsets.csv` in the current working dir.

**Comment by dneise** [comment]
I think we should pull this into the master in order to start putting the calibration application into the EventGetter.

**Comment by Froehliche-Kernschmelze** [comment]
Thanks for your feedback / help , guys! The code is in principal ready to be pulled. Still working on minor issues though.

**Pull request merged**